fulltext/extract.py

```diff
@@ -1,57 +1,252 @@
+
+    try:
+        if id_type == SupportedBuckets.ARXIV:
+            pdf_path = canonical.retrieve(identifier)
+        elif id_type == SupportedBuckets.SUBMISSION:
+            pdf_path, owner = compilations.retrieve(identifier, token)
+        else:
+            RuntimeError(f'Unsupported identifier: {identifier} ({id_type})')
+        extraction = storage.retrieve(identifier, version, bucket=id_type,
+                                      meta_only=True)
+        content = do_extraction(pdf_path)
+        assert content is not None
     except Exception as e:
-        logger.error('Failed to process %s: %s' % (document_id, e))
+        logger.error('Failed to process %s: %s', identifier, e)
+        storage.store(extraction.copy(status=Extraction.Status.FAILED,
+                                      ended=datetime.now(UTC),
+                                      exception=str(e)))
         raise e
-    return {
-        'document_id': document_id,
-    }
-
+    finally:
+        os.remove(pdf_path)    # Cleanup.
 
-extract_fulltext.async_result = AsyncResult
+    extraction = extraction.copy(status=Extraction.Status.SUCCEEDED,
+                                 ended=datetime.now(UTC),
+                                 content=content)
+    storage.store(extraction, 'plain')
+    extraction = extraction.copy(content=psv.normalize_text_psv(content))
+    storage.store(extraction, 'psv')
+    result = extraction.to_dict()
+    result.pop('content')
+    return result
 
 
 @after_task_publish.connect
-def update_sent_state(sender=None, headers=None, body=None, **kwargs):
+def update_sent_state(sender: Optional[str] = None,
+                      headers: Optional[Dict[str, str]] = None,
+                      body: Any = None, **kwargs: Any) -> None:
     """Set state to SENT, so that we can tell whether a task exists."""
-    task = current_app.tasks.get(sender)
-    backend = task.backend if task else current_app.backend
-    backend.store_result(headers['id'], None, "SENT")
+    task = celery_app.tasks.get(sender)
+    backend = task.backend if task else celery_app.backend
+    if headers is not None:
+        backend.store_result(headers['id'], None, "SENT")
+
+
+def do_extraction(filename: str, cleanup: bool = False,
+                  image: Optional[str] = None) -> str:
+    """
+    Extract fulltext from the PDF represented by ``filehandle``.
+
+    Parameters
+    ----------
+    filename : str
+
+    Returns
+    -------
+    str
+        Raw XML response from FullText.
+
+    """
+    logger.info('Attempting text extraction for %s', filename)
+    start_time = datetime.now()
+
+    # This is the path in this container/env where PDFs are stored.
+    workdir = current_app.config['WORKDIR']
+    # This is the path on the Docker host that should be mapped into the
+    # extractor container at /pdf. This is the same volume that should be
+    # mounted at ``workdir`` in this container/env.
+    mountdir = current_app.config['MOUNTDIR']
```

Comment:
I don't see `MOUNTDIR` being set anywhere in the demo service cluster. Does it need to be added to `Dockerfile-worker` or `Docker-compose.yml`?

Reply:
Nevermind; I see that I didn't have the latest changes to `fulltext/config.py` which sets a default for `MOUNT_DIR`.

Reply:
Actually there is something not quite right there. For some reason I am constantly tripping myself up on the volumes.

It should be:

``` 
                      | <-- {workdir} (worker)
[working volume] <--- |
                      | <-- {mountdir} (dind) <-- /pdfs (extractor)
```

Will fix that up

Reply:
For now you can just set ``MOUNTDIR: "/pdfs"`` in your docker-compose (for both API and worker, to be safe)

Reply:
Ok, fixed in  284bd3d